components/scheduler/executors.go

```diff
@@ -23,41 +23,454 @@
 	})
 }
+
+func (e taskExecutor) executeSchedulerRunTask(
+	ctx context.Context,
+	env hsm.Environment,
+	ref hsm.Ref,
+	task SchedulerActivateTask,
+) error {
+	ns, err := e.NamespaceRegistry.GetNamespaceByID(namespace.ID(ref.WorkflowKey.NamespaceID))
+	if err != nil {
+		return fmt.Errorf("failed to get namespace by ID: %w", err)
+	}
+	ctx, cancel := context.WithTimeout(ctx, e.Config.ExecutionTimeout(ref.WorkflowKey.NamespaceID))
+	defer cancel()
+
+	var s Scheduler
+	err = env.Access(ctx, ref, hsm.AccessRead, func(node *hsm.Node) error {
+		if err := node.CheckRunning(); err != nil {
+			return err
+		}
+		sched, err := hsm.MachineData[*Scheduler](node)
+		if err != nil {
+			return err
+		}
+		// Ensure that this is copy, so later (stateful) processing does not update state machine without protection
+		s = *sched
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	// Update in case namespace was renamed since creation
+	s.Args.State.Namespace = ns.Name().String()
+	tweakables := e.Config.Tweakables(s.Args.State.NamespaceId)
+
+	if s.cspec == nil {
+		cspec, err := e.SpecBuilder.NewCompiledSpec(s.Args.Schedule.Spec)
+		if err != nil {
+			if e.Logger != nil {
+				e.Logger.Error("Invalid schedule", tag.Error(err))
+			}
+			s.cspec = nil
+			return err
+		}
+		s.cspec = cspec
+	}
+
+	if s.Args.State.LastProcessedTime == nil {
+		// log these as json since it's more readable than the Go representation
+		specJson, _ := protojson.Marshal(s.Args.Schedule.Spec)
+		policiesJson, _ := protojson.Marshal(s.Args.Schedule.Policies)
+		e.Logger.Info("Starting schedule", tag.NewStringTag("spec", string(specJson)), tag.NewStringTag("policies", string(policiesJson)))
+		s.Args.State.LastProcessedTime = timestamppb.Now()
+		s.Args.State.ConflictToken = scheduler.InitialConflictToken
+		s.Args.Info.CreateTime = s.Args.State.LastProcessedTime
+	}
+
+	t1 := timestamp.TimeValue(s.Args.State.LastProcessedTime)
+	t2 := time.Now()
+	if t2.Before(t1) {
+		e.Logger.Warn("Time went backwards", tag.NewStringTag("time", t1.String()), tag.NewStringTag("time", t2.String()))
+		t2 = t1
+	}
+	nextWakeup, lastAction := e.processTimeRange(
+		&s,
+		&tweakables,
+		t1, t2,
+		// resolve this to the schedule's policy as late as possible
+		enumspb.SCHEDULE_OVERLAP_POLICY_UNSPECIFIED,
+		false,
+		nil,
+	)
+
+	s.Args.State.LastProcessedTime = timestamppb.New(lastAction)
+	s.NextInvocationTime = timestamppb.New(nextWakeup)
+	// process backfills if we have any too
+	e.processBackfills(&s, &tweakables)
+
+	// try starting workflows in the buffer
+	// nolint:revive
+	for {
+		hasNext, err := e.processBuffer(ctx, env, ref, &s)
+		if err != nil {
+			return err
+		}
+		if !hasNext {
+			break
+		}
+	}
+
+	return env.Access(ctx, ref, hsm.AccessWrite, func(node *hsm.Node) error {
+		sPtr, err := hsm.MachineData[*Scheduler](node)
+		if err != nil {
+			return err
+		}
+		// Copy outside scheduler, which has been updated, to the state machine
+		*sPtr = s
```

Comment:
Does env not check this automatically?

Either way, I added a simple check for whether the state has changed since last updated. We should probably chat about whether this is something we want to build into the framework instead of relying on users .

Reply:
This isn't ideal but let's just move on, we can refactor later.